### ISSUE TITLE
More hotkeys / keyboard shortcuts

### ISSUE THREAD

**Issue by richvdh** [element-hq/element-web#3232]
We have a few requests for extra hotkeys, but I thought a single issue would be useful to help track them.

So far we have:

Working:
* ctrl-d / ctrl-e in voip calls to mute/unmute mic / camera
* ctrl-k to jump to named room
* up/down to navigate within room list
* tab/shift-tab to navigate focus as normal (other than in MessageComposer, where it tab-completes)
* page-up/page-down to navigate within the main scrollpanel
* ctrl-home/ctrl-end to jump to beginning or end of main scrollpanel

Broken:
- [x] alt-up/down to switch rooms, though it is broken (https://github.com/vector-im/riot-web/issues/49)
- [ ] ctrl-shift-number to jump between the first 9 rooms, though it is broken (as it collides with switching tabs or something on some browsers)
- [ ] page up/down in the room directory (#1803)

We should add hotkeys for
* [x] Next unread room (https://github.com/vector-im/riot-web/issues/301)
* [ ] Message search
* [ ] Send an invite

This is related to https://github.com/vector-im/riot-web/issues/302, but I think that's more about tab indexes and making it possible to "click" buttons without a mouse, while this is all about hotkeys that don't require particular elements to be focussed.

**Comment by albuic** [comment]
+1 pour le #301 (Next unread room)

**Comment by frankiebee** [comment]
- ctrl-k to jump to named room
does not work in macOsx native app or on chrome 

- up/down 
 up/down does not seem to work

**Comment by t3chguy** [comment]
@frankiebee I believe on Mac it is Cmd+k

**Comment by frankiebee** [comment]
that does not work either

**Comment by anton-ryzhov** [comment]
What about pretty standard feature — `Enter` for newline, `Ctrl-Enter` for send?

And another hotkey-related request. I's really useful to close all chats to keep all messages unread. Currently it can be done with `Home` button in the bottom of left menu.
Will it be nice to have hotkey for this? For example Telegram uses `ESC` for this — very convenient and expectable.

**Comment by t3chguy** [comment]
@anton-ryzhov more commonly its `Shift-Enter` for newline and `Enter` for send.

**Comment by anton-ryzhov** [comment]
@t3chguy it doesn't really matter if it is configurable by user)

**Comment by t3chguy** [comment]
@anton-ryzhov it's down to your browser how it handles input boxes

**Comment by aperezdc** [comment]
The `Ctrl+K` shortcut does _not_ work for me either in WebKitGTK+ based browsers (or in [Revolt](/aperezdc/revolt), which also uses it).

**Comment by t3chguy** [comment]
`Ctrl+K` was removed while a PR undergoes testing, it will be added back when it is tested. https://github.com/matrix-org/matrix-react-sdk/pull/895

**Comment by anton-ryzhov** [comment]
@t3chguy 
> it's down to your browser how it handles input boxes

What do you mean? It is a <textarea> , it doesn't send form by default in browsers. But in `riot` it does — this is `riot`s behaviour, no?

**Comment by t3chguy** [comment]
Depends, it can also be a contenteditable div.
`<div aria-describedby="placeholder-1soc8" class="public-DraftEditor-content" contenteditable="true" role="textbox" spellcheck="true" style="outline: none; white-space: pre-wrap; word-wrap: break-word;"><div data-contents="true"><div data-block="true" data-editor="1soc8" data-offset-key="2cmcm-0-0"><div data-offset-key="2cmcm-0-0" class="public-DraftStyleDefault-block public-DraftStyleDefault-ltr"><span data-offset-key="2cmcm-0-0"><br data-text="true"></span></div></div></div></div>`

**Comment by tanius** [comment]
Adding to the list :-)

- [ ] `Ctrl+Home` / `Ctrl+End` to jump to beginning or end of multiline text input in the Composer. "Like in normal editors."
- [ ] `Ctrl+PageUp` / `Ctrl+PageDown` to jump to beginning / end of the main scrollpanel. Currently assigned to `Ctrl+Home` / `Ctrl+End` but needs a new binding if implementing the task just above. `PageUp` / `PageDown` currently scrolls through the main scrollpanel, so the new proposal seems a good fit to me.

**Comment by Alpherie** [comment]
`Enter` for newline, `Ctrl-Enter` for send would definitely be nice, I prefer this way.

**Comment by coreinsane** [comment]
It would be nice to Bind the shortcuts for yourself, for example on mouse buttons, and so on.

**Comment by t3chguy** [comment]
@coreinsane I presume you mean on electron app as mouse buttons tend to be bound at browser level

**Comment by coreinsane** [comment]
Yes you got the point, im using windows desktop client.

**Comment by coreinsane** [comment]
Is there any improvement on windows desktop client hotkey features? Like customisable hotkeys, can be bound on mouse, etc.

**Comment by pvagner** [comment]
It would be nice to add some more hotkeys related to the infinite scrolling functionality of the room timeline because hotkeys such as ctrl+home, ctrl+end, pageup and page down are usually clashing with standard browser keyboard commands when caret browsing is turned on (e.g. by pressing F7 in Firefox). I would like to have at least shortcuts for scrolling to the bottom and then scroll to prev / next page.

**Comment by eatonphil** [comment]
Are Riot hotkeys documented somewhere permanently?

**Comment by ptman** [comment]
@eatonphil I started documenting them here https://github.com/vector-im/riot-web/pull/7165

**Comment by madduck** [comment]
All of these shortcuts should be configurable. Please do not create a jumble of hardcoded key combos.

**Comment by colans** [comment]
> tab/shift-tab to navigate focus as normal (other than in MessageComposer, where it tab-completes)

How does one jump out of MessageComposer then?  Once you're there, there doesn't appear to be a way out.

**Comment by t3chguy** [comment]
> How does one jump out of MessageComposer then? Once you're there, there doesn't appear to be a way out.

The plan is to stop trapping tab there

**Comment by foresto** [comment]
How about a shortcut for jumping to the first unread message?

**Comment by ptman** [comment]
@foresto #301

**Comment by foresto** [comment]
> #301

That one is for jumping to another room with unread messages. I'm talking about jumping to the first unread message in the current room.

When I change to another browser tab for a while, or walk away from my computer and then come back, there are sometimes a bunch of new messages in current room. It would be nice not to have to manually scroll back and read every message in reverse just to find the point where I left off.

**Comment by ptman** [comment]
@foresto sorry, I should've read more carefully

**Comment by jaywink** [comment]
I literally daily try to do `ctrl+tab` to switch between two rooms, like I can switch between two tabs in my Firefox. Of course it wouldn't really work for riot-web but would be awesome for the desktop version :thinking:

**Comment by t3chguy** [comment]
Right now that's alt up-down, we could add that for electron. Upvote the comment if desired.

**Comment by jaywink** [comment]
> Right now that's alt up-down

That iterates the room list up/down though, which is useful but not entirely the same thing. Some apps like Chrome I believe also iterate tabs on ctrl-tab. But often the behaviour is to cycle between the last used ones, so that you can for example switch between two rooms back and forth, even if your room list is 100. This is common I believe in many text editors but Firefox also has it, though I believe it needs to be separately enabled AFAICR.

Of course alt up/down works if the rooms one is cycling are relatively close to each other in the room list.

**Comment by t3chguy** [comment]
That's #6384 then

**Comment by foresto** [comment]
>     * ctrl-k to jump to named room does not work in macOsx native app or on chrome

Ctrl-K does not work on Riot desktop for linux, either.

**Comment by jaywink** [comment]
> Ctrl-K does not work on Riot desktop for linux, either.

@foresto It definitely should work, the comment you referred to is from 2017. Running Riot Desktop on Ubuntu myself.

Possibly something is conflicting with it in your setup? I would suggest logging a new issue with more details on linux desktop flavour, window manager, and such? And obviously checking first if you have some custom shortcuts added that conflict with it.

**Comment by t3chguy** [comment]
@foresto 

> Ctrl-K does not work on Riot desktop for linux, either.

Do you have the left panel collapsed, is the Explore button visible?

**Comment by foresto** [comment]
@t3chguy 
> Do you have the left panel collapsed

Yes, I have it collapsed. I tried it just now with the left panel expanded, and Ctrl-K worked in that case. (Doesn't do me any good, though, since an expanded left panel wastes too much space.)

**Comment by t3chguy** [comment]
Now it'll work even when collapsed

**Comment by foresto** [comment]
Confirmed: Ctrl-K now works when the left panel is collapsed. (Thanks, @t3chguy !) However, it also expands the panel and leaves it expanded, dominating the window and pushing the chat area into a narrow column. I end up having to pick up the mouse to fix this, so it ends up no more helpful than it was when it didn't work.

However, alt-up/down seems to work now, which is handy. :)

**Comment by tredondo** [comment]
Ctrl+F would be great for search. Lots of muscle memory from browser etc.

**Comment by t3chguy** [comment]
That's #2764

**Comment by Darksider3** [comment]
I'd propose remapping CTRL+K to markdown links and "current room search" as ctrl+shift+f(which would IMHO make more sense when "serach **in** room" is CTRL + f ). 
My muscle memory from other editors(github, tyora, obsidian, marktext) is pretty useless in that regard currently.

**Comment by t3chguy** [comment]
`CTRL+K` is already jump to room in other chat apps so for consistency it makes sense to keep it as such, see Discord and Slack

**Comment by Darksider3** [comment]
Since you're mentioning Slack and Discord, why not CTRL+shift+U? :stuck_out_tongue:

**Comment by t3chguy** [comment]
Riot:
![image](https://user-images.githubusercontent.com/2403652/84019056-b4a36c00-a978-11ea-98e3-4d743c9e8520.png)

Discord:
![image](https://user-images.githubusercontent.com/2403652/84019074-ba00b680-a978-11ea-82ed-433731b4e9c4.png)

Slack, I don't see Ctrl Shift U in https://slack.com/intl/en-gb/help/articles/201374536-Slack-keyboard-shortcuts#windows-linux-6

**Comment by Darksider3** [comment]
Well seems like they've removed it in february this year, according to [this answer on SO](https://stackoverflow.com/questions/39105038/create-hyperlink-in-slack)

**Comment by hex-m** [comment]
> Since you're mentioning Slack and Discord, why not CTRL+shift+U? stuck_out_tongue

At least on GNOME that's already taken.

> To enter a character by its code point, press Ctrl+Shift+U, then type the four-character code and press Space or Enter.

([src](https://help.gnome.org/users/gnome-help/stable/tips-specialchars.html.en#ctrlshiftu))

**Comment by HarHarLinks** [comment]
Add keys to open Threads, Pins, Notifications panels.

**Comment by lech** [comment]
> All of these shortcuts should be configurable. Please do not create a jumble of hardcoded key combos.

This would be seriously helpful as rather than CTRL+Y to redo CTRL+SHIFT+Z is an alternate key combination to perform the same thing in most modern applications depending on the platform.

In addition in one of the latest updates the "Search" icon is now just displaying as "CTRL+K" when the sidebar is collapsed 😞

**Comment by presto8** [comment]
What is the policy or thinking of multiple shortcuts for the same action? I prefer to keep my fingers on the home row which makes accessing the arrow keys non-optimal. Shortcuts like Alt+Shift+Down for next unread room are very awkward to type for people used to vim/emacs type bindings.

Ctrl+K is wonderful for jump to room. Would it be possible to get Ctrl+J to "jump" to next unread conversation (Ctrl+Shift+J cycles back in the list.)

Thanks!

**Comment by vertigo220** [comment]
It was already mentioned, but ctrl+tab is used pretty universally to switch back and forth between tabs in all browsers, Notepad++, PDF programs, etc, and I definitely think it should be used here for switching back and forth between rooms. The best implementation of this would be to show a drop-down list of rooms when it's pressed, and pressing it once just switches back to the last room whereas holding the ctrl key down and pressing tab repeatedly moves through the list to choose a different room. See Notepad++ for an example.

Hotkeys should also be customizable if at all possible, so people can decide for themselves which ones they want and don't want and what key combinations they want to use for them. That would solve a lot of issues of people asking for specific hotkeys or asking if changes can be made to them.

**Comment by HarHarLinks** [comment]
> ctrl+tab is used pretty universally to switch back and forth between tabs in all browsers

exactly the reason that element can't adopt this shortcut: it (at least E web) runs in the browser.

**Comment by vertigo220** [comment]
> exactly the reason that element can't adopt this shortcut: it (at least E web) runs in the browser.

Fair enough. Too much work to have separate shortcuts in browser vs desktop I take it? Because it would work in desktop, but I can see that might be more work having two separate versions to maintain.

**Comment by Miirek** [comment]
Please add a ⌘-0  (CMD+zero) app shortcut to call main conversation window. I cannot do it by myself because menu field is changing with room title.

**Comment by t3chguy** [comment]
Cmd-0 is a pre-existing browser shortcut for reset zoom.

<img width="258" alt="image" src="https://github.com/element-hq/element-web/assets/2403652/9ca588dd-9f03-4100-b757-0f14eb084340">

**Comment by Miirek** [comment]
Ah, I see. It also depends on localization. In my case it is **⌘-é**  So I suggest ⌘-1 or **something** else - the need to pick mouse, find minimized window and click is quite frustrating and time consuming - when you are used to use mostly keyboard ;-) Most mac apps uses some kind of a shorcut to view main app window (mail, messages, even MS Skype does ...)

**Comment by t3chguy** [comment]
You just want it to open the app? You can normally do this in your OS. https://support.apple.com/en-gb/guide/mac-help/mchlp2271/mac#:~:text=On%20your%20Mac%2C%20choose%20Apple,Keyboard%20Shortcuts%20on%20the%20right.&text=Select%20App%20Shortcuts%20on%20the,specific%20app%20or%20All%20Applications.

Element is a single instance app so it'll open the existing window if the app is already running.

**Comment by Miirek** [comment]
@t3chguy - thanks for quick response ... sure, application is already opened - but window is minimized. When I want to focus to the main window I use ⌘-tab to switch between apps until its Element and _sometimes_ window pops up (it was only in background, behind other windows) but very often nothing shows. This means window is minimized and must be clicked. And from some other apps I'm used to push ⌘-0 or ⌘-1 or ⌘-⇧-0 or something else to call main window. It seems to me, that adding simple shortcut to the window is easiest way to achieve it ;-) Sometimes it can be done without developers assistance, but its not this case - menu item with main window has mutating title, so I cant assign a hotkey from outside.
Adding global hotkey is not always working - one has to choose an unusual combination which doesnt already exist ...   
 Anyway, thanks for your time.

**Comment by t3chguy** [comment]
Why can you not assign the global hotkey to just run the Element.app executable rather than trying to do magic dependant on window titles?

**Comment by Miirek** [comment]
I have one - ⌘-⇧-^-E - this is a combination which works for me and is not in conflict with some other app. I'm just suggesting to follow a common (IMHO) behavior ...

**Comment by t3chguy** [comment]
Adding such a global shortcut would be blocked on the ability to have customisable keybinds otherwise it'd conflict for people, we avoid global shortcuts for that reason

**Comment by Miirek** [comment]
I don't want global - I don't like it, it can cause problems, conflicts, unintended behavior and so. I prefer local, in-app shortcut.

**Comment by t3chguy** [comment]
Sorry, how would a non-global shortcut help you bring the app to foreground given it wouldn't be receiving keys if it wasn't already in foreground focus?

**Comment by Miirek** [comment]
As described earlier - I switch between apps using ⌘-Tab to focus to Element and then I miss a shortcut to call minimized window. Thats all.
![image](https://github.com/element-hq/element-web/assets/1177720/64b12372-96c5-4075-99a0-b1c013b3266b)

**Comment by t3chguy** [comment]
Sure, but if its minimized then a local shortcut won't be picked up, only a global one would, as the window is not in focus.

**Comment by Miirek** [comment]
Hmmm, I dont agree, try with, for example, Messages app - minimize app window and press ⌘-0 and you will see window coming to the screen. Even when the window is closed. Of course you must have Messages  active and focused (in top bar)

**Comment by t3chguy** [comment]
It may well be that way in Messages, but AFAICT it isn't that way for Electron apps

**Comment by richvdh** [comment]
This issue appears pretty outdated. At least some of the things under the "broken" and "things we should add" lists are now fixed. In any case, it's never going to be "done".

As the original reporter, I'm not really interested in the debate about the particular intricacies of global shortcuts in OSX. I'd suggest if you want to continue the conversation you open a specific issue for it. Meanwhile, I'm closing this one.

**Comment by CyberShadow** [comment]
> I'm not really interested in the debate about the particular intricacies of global shortcuts in OSX

@richvdh I think you can click "Unsubscribe" and stop receiving notifications even if you're the author.

FWIW / FYI, I see @t3chguy edited the issue to track the status of some items, so perhaps it might not be useful to you but it might be useful for the project? <sup>(I don't have an opinion either way)</sup>
